mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -0,0 +1,422 @@
+    } else {
+        // if the request size is smaller than the block size, just return the block size
+        if preferred_size < block_size as usize {
+            block_size as usize
+        } else {
+            // if it exceeds block boundaries, trim it to the block boundaries
+            let request_boundary = offset + preferred_size as u64;
+            let remainder = request_boundary % block_size;
+            if remainder != 0 {
+                preferred_size + (block_size - remainder) as usize
+            } else {
+                preferred_size
+            }
+        }
+    };
+    RequestRange::new(object_size, offset, size)
+}
+
+#[derive(Debug)]
+struct CachingRequest<Client: ObjectClient, Cache> {
+    client: Client,
+    cache: Arc<Cache>,
+    bucket: String,
+    cache_key: CacheKey,
+    part_queue_producer: PartQueueProducer<Client::ClientError>,
+}
+
+impl<Client, Cache> CachingRequest<Client, Cache>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+    Cache: DataCache + Send + Sync,
+{
+    fn new(
+        client: Client,
+        cache: Arc<Cache>,
+        bucket: String,
+        key: String,
+        etag: ETag,
+        part_queue_producer: PartQueueProducer<Client::ClientError>,
+    ) -> Self {
+        let cache_key = CacheKey { s3_key: key, etag };
+        Self {
+            client,
+            cache,
+            bucket,
+            cache_key,
+            part_queue_producer,
+        }
+    }
+
+    async fn get_from_cache(&self, range: RequestRange) {
+        let key = &self.cache_key.s3_key;
+        let block_size = self.cache.block_size();
+        let block_range = self.block_indices_for_byte_range(&range);
+
+        // TODO: consider starting GetObject requests pre-emptively if cache blocks are missing
+        for block_index in block_range.clone() {
+            match self.cache.get_block(&self.cache_key, block_index) {
+                Ok(Some(block)) => {
+                    trace!(?key, ?range, block_index, "cache hit");
+                    let part = self.make_part(block, block_index, &range);
+                    metrics::counter!("cache.total_bytes", part.len() as u64, "type" => "read");
+                    self.part_queue_producer.push(Ok(part));
+                }
+                Ok(None) => {
+                    trace!(?key, ?range, block_index, "cache miss - no data for block");
+                    return self
+                        .get_from_client(range.trim_start(block_index * block_size), block_index..block_range.end)
+                        .await;
+                }
+                Err(error) => {
+                    trace!(?key, ?range, block_index, ?error, "error reading block from cache");
+                    return self
+                        .get_from_client(range.trim_start(block_index * block_size), block_index..block_range.end)
+                        .await;
+                }
+            }
+        }
```

Comment:
Maybe this version makes the logic a little clearer? The point being that we fail out immediately if we fail to read a cache block or if a block isn't cached.
```rust
// TODO: consider starting GetObject requests pre-emptively if cache blocks are missing
for block_index in block_range.clone() {
    match self.cache.get_block(&self.cache_key, block_index) {
        Ok(Some(block)) => {
            trace!(?key, ?range, block_index, "cache hit");
            let part = self.make_part(block, block_index, &range);
            metrics::counter!("cache.total_bytes", part.len() as u64, "type" => "read");
            self.part_queue_producer.push(Ok(part));
            continue;
        }
        Ok(None) => trace!(?key, ?range, block_index, "cache miss - no data for block"),
        Err(error) => trace!(?key, ?range, block_index, ?error, "error reading block from cache"),
    }

    // If a block is uncached or reading it fails, we don't try to recover and just fall back to S3
    // for the entire stream
    return self
        .get_from_client(range.trim_start(block_index * block_size), block_index..block_range.end)
        .await;
}
```

Reply:
also, probably `error!` in the error case, it's not something we expect in normal operation?

Reply:
Yes. I also had gone back and forth on this. Another option to consider - maybe later - is for `get_block` to return an error on cache miss.